lib/jsonapi/utils/underscore.ex

```diff
@@ -0,0 +1,17 @@
+defmodule JSONAPI.Utils.Underscore do
+  def underscore?, do: Application.get_env(:jsonapi, :underscore_to_dash, false)
+
+  # TODO Figure out the nesting of fields
+  def underscore(data) do
+    data
+    |> Enum.map(fn {key, value} -> {underscore_key(key), value} end)
+    |> Enum.into(%{})
+  end
+
+  def underscore_key(key) do
```

Comment:
Perhaps we could refactor out the explicit `underscore_key/1` and rely on pattern matching instead, that might help us solve nesting:

```elixir
defmodule JSONAPI.Utils.Underscore do
  def underscore?, do: Application.get_env(:jsonapi, :underscore_to_dash, false)

  def underscore(value) when is_atom(value) do
    value
    |> to_string
    |> underscore
  end

  def underscore(value) when is_bitstring(value) do
    value
    |> String.replace("_", "-")
    |> String.to_atom
  end

  def underscore(value) when is_map(value) do
    value
    |> Enum.map(&underscore/1)
    |> Enum.into(%{})
  end

  def underscore({key, value}) do
    {underscore(key), underscore(value)}
  end
end
```

Reply:
Sounds good!  I had to change the final clause as underscoring map string values caused some issues.